# HG changeset patch
# User alta88@fixall.com
# Date 1703178633 -3600
# Parent  df58e19c81cb1323ccacd7b70fcc35c5c73fca4d
Bug 323067 - No indicator of currently sorted column when column header is icon only

Add a sort indicator to icon columns and add a secondary sort indicator to all columns.

Authored by Alta88 and contributed to the Betterbird Project.
See https://github.com/Betterbird/thunderbird-patches/blob/main/LICENSE for license details.
All uses require attribution to the Author.

diff --git a/mail/base/content/about3Pane.js b/mail/base/content/about3Pane.js
--- a/mail/base/content/about3Pane.js
+++ b/mail/base/content/about3Pane.js
@@ -5622,16 +5622,40 @@ var threadPane = {
       .querySelector(".sorting")
       ?.classList.remove("sorting", "ascending", "descending");
     this.treeTable
       .querySelector(`#${column} button`)
       ?.classList.add(
         "sorting",
         gViewWrapper.isSortedAscending ? "ascending" : "descending"
       );
+
+    let secondaryColumn = sortController.convertSortTypeToColumnID(
+      gViewWrapper.dbView.secondarySortType
+    );
+    this.treeTable
+      .querySelector(".sortingSecondary")
+      ?.classList.remove("sortingSecondary", "ascending", "descending");
+
+    // Don't add secondary sort indicator in these cases:
+    // "Order received" is unique, dates will be "mostly" unique,
+    // no need to clutter the UI with unnecessary secondary indicators.
+    if (["idCol", "dateCol", "receivedCol"].includes(column)) {
+      return;
+    }
+
+    this.treeTable
+      .querySelector(`#${secondaryColumn} button`)
+      ?.classList.add(
+        "sortingSecondary",
+        gViewWrapper.dbView.secondarySortOrder ==
+          Ci.nsMsgViewSortOrder.ascending
+            ? "ascending"
+            : "descending"
+      );
   },
 
   /**
    * Prompt the user to confirm applying the current columns state to the chosen
    * folder and its children.
    *
    * @param {nsIMsgFolder} folder - The chosen message folder.
    * @param {boolean} [useChildren=false] - If the requested action should be
diff --git a/mail/themes/shared/mail/tree-listbox.css b/mail/themes/shared/mail/tree-listbox.css
--- a/mail/themes/shared/mail/tree-listbox.css
+++ b/mail/themes/shared/mail/tree-listbox.css
@@ -6,40 +6,48 @@
   --listbox-selected-outline: none;
   --tree-header-table-height: 30px;
   --tree-header-cell-padding: 6px;
   --tree-header-cell-indent: 0;
   --list-item-min-height: 26px;
   --list-item-buttons-size: 22px;
   --table-header-cell-icon-width: 24px;
   --table-header-cell-icon-min-width: 24px;
+  --table-header-cell-sortingSecondary-position: center,13px;
+  --table-header-cell-icon-sort-indicator-position: 11px;
   --table-row-cell-img-margin-inline-start: 6px;
   --table-row-cell-button-margin-inline-start: 4px;
 }
 
 :root[uidensity="compact"] {
   --tree-header-table-height: 27px;
   --tree-header-cell-padding: 3px;
   --tree-header-cell-indent: 3px;
   --list-item-min-height: 18px;
   --list-item-buttons-size: 18px;
   --table-header-cell-icon-width: 18px;
   --table-header-cell-icon-min-width: 18px;
   --table-header-cell-icon-button-padding: 0;
+  --table-header-cell-sortingSecondary-position: center,11px;
+  --table-header-cell-icon-sort-indicator-position: 10px;
+  --table-header-cell-icon-sort-indicator-size: 7px;
+  --table-header-cell-icon-sortingSecondary-position: center,13px;
   --table-row-cell-img-margin-inline-start: 3px;
   --table-row-cell-button-margin-inline-start: 1px;
 }
 
 :root[uidensity="touch"] {
   --tree-header-table-height: 36px;
   --tree-header-cell-padding: 9px;
   --tree-header-cell-indent: -3px;
   --list-item-min-height: 32px;
   --table-header-cell-icon-width: 33px;
   --table-header-cell-icon-min-width: 33px;
+  --table-header-cell-sortingSecondary-position: center,16px;
+  --table-header-cell-icon-sort-indicator-position: 18px;
   --table-row-cell-img-margin-inline-start: 11px;
   --table-row-cell-button-margin-inline-start: 9px;
 }
 
 @media not (prefers-contrast) {
   :root {
     --listbox-color: var(--color-gray-80);
     --listbox-selected-bg: var(--color-gray-20);
@@ -364,29 +372,53 @@ th[data-resizable="false"] .tree-table-c
   background-image: var(--icon-nav-up-sm);
   padding-inline-end: 18px;
 }
 
 .tree-table-cell-container button.sorting.descending {
   background-image: var(--icon-nav-down-sm);
 }
 
+.tree-table-cell-container button.sortingSecondary {
+  background-image: var(--icon-nav-up-sm), var(--icon-nav-up-sm);
+  background-position-y: var(--table-header-cell-sortingSecondary-position);
+}
+
+.tree-table-cell-container button.sortingSecondary.descending {
+  background-image: var(--icon-nav-down-sm), var(--icon-nav-down-sm);
+}
+
 @media (-moz-platform: linux) {
   .tree-table-cell-container button.sorting {
     background-image: var(--icon-nav-down-sm);
   }
 
   .tree-table-cell-container button.sorting.descending {
     background-image: var(--icon-nav-up-sm);
   }
+
+  .tree-table-cell-container button.sortingSecondary {
+    background-image: var(--icon-nav-down-sm), var(--icon-nav-down-sm);
+  }
+
+  .tree-table-cell-container button.sortingSecondary.descending {
+    background-image: var(--icon-nav-up-sm), var(--icon-nav-up-sm);
+  }
 }
 
-th[data-type="icon"] .tree-table-cell-container button.sorting {
-  background-image: none;
-  padding-inline-end: var(--tree-header-cell-padding);
+th[data-type="icon"] .tree-table-cell-container button.sorting,
+th[data-type="icon"] .tree-table-cell-container button.sortingSecondary {
+  padding-inline: 0 12px;
+  background-size: var(--table-header-cell-icon-sort-indicator-size);
+  background-position-x: var(--table-header-cell-icon-sort-indicator-position);
+}
+
+th[data-type="icon"] .tree-table-cell-container button.sortingSecondary {
+  background-position-y: var(--table-header-cell-icon-sortingSecondary-position,
+                         var(--table-header-cell-sortingSecondary-position));
 }
 
 th[is="tree-view-table-column-picker"] {
   position: relative;
   width: 25px;
   min-width: 25px;
   padding: 0;
   background-color: var(--tree-view-bg);
